File: nl.utwente.ewi.fmt.uppaalsmc.urpal/main/java/nl/utwente/ewi/fmt/uppaalSMC/urpal/properties/UnusedVariablesProperty.java
package nl.utwente.ewi.fmt.uppaalSMC.urpal.properties;

import java.awt.Color;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import javax.swing.BoxLayout;
import javax.swing.JLabel;
import javax.swing.JPanel;

import org.eclipse.emf.ecore.EObject;
import org.eclipse.emf.ecore.util.EcoreUtil;
import org.muml.uppaal.core.NamedElement;
import org.muml.uppaal.declarations.Variable;

import com.uppaal.model.core2.Document;
import com.uppaal.model.system.UppaalSystem;

import nl.utwente.ewi.fmt.uppaalSMC.NSTA;

@SanityCheck(name = "Check for unused elements")
public class UnusedVariablesProperty extends AbstractProperty {

	@Override
	public void doCheck(NSTA nsta, Document doc, UppaalSystem sys, Consumer<SanityCheckResult> cb) {
		List<Variable> unusedVars = new ArrayList<>();
		nsta.eAllContents().forEachRemaining(obj -> {
			if (obj instanceof Variable) {
				if (EcoreUtil.UsageCrossReferencer.find(obj, nsta).isEmpty()) {
					unusedVars.add((Variable) obj);
				}
			}
		});
		List<String> qualifiedNames = unusedVars.stream().map(v -> {
			StringBuilder sb = new StringBuilder(v.getName().replace("nta.", "") + "(" + v.eContainer().getClass().getSimpleName()
					.replace("Declaration", "")
					.replace("Impl", "") + ")");
			EObject obj = v;
			while ((obj = obj.eContainer()) != null) {
				if (obj instanceof NamedElement) {
					sb.insert(0, ((NamedElement) obj).getName() + ".");
				}
			}
			return sb.toString();
		}).collect(Collectors.toList());
		cb.accept(new SanityCheckResult() {

			@Override
			public void write(PrintStream out, PrintStream err) {
				if (qualifiedNames.isEmpty()) {
					out.println("No unused variables found");
				} else {
					err.println("Unused variables found: ");
					qualifiedNames.forEach(err::println);
				}
			}

			@Override
			public JPanel toPanel() {
				if (qualifiedNames.isEmpty()) {
					JPanel p = new JPanel();
					p.setLayout(new BoxLayout(p, BoxLayout.Y_AXIS));
					p.add(new JLabel("All variables used!"));
					return p;
				} else {
					JPanel p = new JPanel();
					p.setLayout(new BoxLayout(p, BoxLayout.Y_AXIS));
					JLabel label = new JLabel("Unused declarations found:");
					label.setForeground(Color.RED);
					p.add(label);
					qualifiedNames.forEach(n -> {
						JLabel locLabel = new JLabel(n);
						locLabel.setForeground(Color.RED);
						p.add(locLabel);
					});

					return p;
				}
			}
		});
	}

}

File: plugin/main/java/nl/utwente/ewi/fmt/uppaalSMC/urpal/properties/UnusedVariablesProperty.kt
package nl.utwente.ewi.fmt.uppaalSMC.urpal.properties

import com.uppaal.model.core2.Document
import com.uppaal.model.system.UppaalSystem
import nl.utwente.ewi.fmt.uppaalSMC.NSTA
import org.eclipse.emf.ecore.EObject
import org.eclipse.emf.ecore.util.EcoreUtil
import org.muml.uppaal.core.NamedElement
import org.muml.uppaal.declarations.Variable
import java.awt.Color
import java.io.PrintStream
import javax.swing.BoxLayout
import javax.swing.JLabel
import javax.swing.JPanel

@SanityCheck(name = "Check for unused elements")
class UnusedVariablesProperty : AbstractProperty() {
	override fun doCheck(nsta: NSTA, doc: Document, sys: UppaalSystem, cb: (SanityCheckResult) -> Unit) {
		val unusedVars = mutableListOf<Variable>()
		nsta.eAllContents().forEachRemaining {
			if (it is Variable) {
				if (EcoreUtil.UsageCrossReferencer.find(it, nsta).isEmpty()) {
					unusedVars.add(it)
				}
			}
		}
		val qualifiedNames = unusedVars.map  {
			val sb = StringBuilder(it.name.replace("nta.", "") + "(" + it.eContainer().javaClass.simpleName
					.replace("Declaration", "")
					.replace("Impl", "") + ")")
			var obj: EObject = it
			generateSequence { obj = obj.eContainer(); obj }
					.filterIsInstance<NamedElement>().forEach { c ->
				sb.insert(0, c.name + ".")
			}
			sb.toString()
		}
		cb(object : SanityCheckResult() {
			override fun write(out: PrintStream, err: PrintStream) {
				if (qualifiedNames.isEmpty()) {
					out.println("No unused variables found")
				} else {
					err.println("Unused variables found: ")
					qualifiedNames.forEach(err::println)
				}
			}

			override fun toPanel(): JPanel {
				val p = JPanel()
				p.layout = BoxLayout(p, BoxLayout.Y_AXIS)
				if (qualifiedNames.isEmpty()) {
					p.add(JLabel("All variables used!"))
				} else {
					val label = JLabel("Unused declarations found:")
					label.foreground = Color.RED
					p.add(label)
					qualifiedNames.forEach {
						val locLabel = JLabel(it)
						locLabel.foreground = Color.RED
						p.add(locLabel)
					}
				}
				return p
			}
		})
	}
}
